### PR TITLE
Fsm state

### PR DESCRIPTION
**Pull request by Frozen**

### CODE REVIEW

File: pkg/state/api.go [hunk 1]
```diff
@@ -110,6 +109,9 @@ type StateInfo interface {
 	// True if state stores additional information in order to provide extended API.
 	ProvidesExtendedApi() (bool, error)
 
+	// Map on readable state. Way to apply multiple operations under same lock
+	MapR(func(StateInfo) (interface{}, error)) (interface{}, error)
```
Comment (alexeykiselev) [inline pkg/state/api.go:113]: The naming is not good... Comment doesn't match with name. And `interface{}` everywhere, is it OK? Do we really need this function? Maybe a few of specific functions will be enough? How many different usages of this `MapR` do we have?

File: pkg/state/batched_storage_test.go [hunk 2]
```diff
@@ -261,7 +261,7 @@ func TestIterators(t *testing.T) {
 	to.testIterator(t, key0, key0Records)
 	to.testIterator(t, key1, key1Records)
 
-	// Add again.
+	// NewConnection again.
```
Comment (alexeykiselev) [inline pkg/state/batched_storage_test.go:264]: Fix renaming

File: pkg/api/app_state.go [hunk 3]
```diff
@@ -1,12 +1,16 @@
...
 
+// TODO Here should be internal message with rollback action
 func (a *App) RollbackToHeight(apiKey string, height proto.Height) error {
-	err := a.checkAuth(apiKey)
-	if err != nil {
-		return err
-	}
-	defer a.state.Mutex().Lock().Unlock()
-	return a.state.RollbackToHeight(height)
+	return errors.New("api method disabled")
```
[the remark this answers is not in the document]
Comment (Frozen) [reply pkg/api/app_state.go:10]: no, just cleanup

File: pkg/node/blocks_applier/blocks_applier.go [hunk 4]
```diff
@@ -101,34 +108,26 @@ func (a *innerBlocksApplier) apply(blocks []*proto.Block) (*proto.Block, proto.H
...
-	lastBlock, _, err := a.inner.apply(blocks)
+func (a *BlocksApplier) Apply(state state.State, blocks []*proto.Block) error {
+	_, _, err := a.inner.apply(state, blocks)
 	if err != nil {
 		return err
 	}
-	if err := maybeEnableExtendedApi(a.state, lastBlock, proto.NewTimestampFromTime(a.tm.Now())); err != nil {
-		panic(fmt.Sprintf("[*] BlockDownloader: MaybeEnableExtendedApi(): %v. Failed to persist address transactions for API after successfully applying valid blocks.", err))
-	}
+	// TODO extended api
```
[the remark this answers is not in the document]
Comment (Frozen) [reply pkg/node/blocks_applier/blocks_applier.go:127]: disabled

**Pull request merged**